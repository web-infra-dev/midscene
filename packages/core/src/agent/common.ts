import type { AbstractPage } from '@/device';
import type { DeviceAction } from '@/index';
import { getMidsceneLocationSchema, z } from '@/index';
import { sleep } from '@/utils';
import type { ElementInfo } from '@midscene/shared/extractor';
import { _keyDefinitions } from '@midscene/shared/us-keyboard-layout';
import { assert } from '@midscene/shared/utils';
import { getKeyCommands } from './ui-utils';

// Generic function to define actions with proper type inference
export const defineAction = <TSchema extends z.ZodType>(
  config: {
    name: string;
    description: string;
    interfaceAlias?: string;
    paramSchema: TSchema;
    call: (param: z.infer<TSchema>) => Promise<void>;
  } & Partial<
    Omit<
      DeviceAction<z.infer<TSchema>>,
      'name' | 'description' | 'interfaceAlias' | 'paramSchema' | 'call'
    >
  >,
): DeviceAction<z.infer<TSchema>> => {
  return config;
};

// Tap
const actionTapParamSchema = z.object({
  locate: getMidsceneLocationSchema().describe('The element to be tapped'),
});
export type ActionTapParam = z.infer<typeof actionTapParamSchema>;

export const defineActionTap = (
  call: (param: ActionTapParam) => Promise<void>,
): DeviceAction<ActionTapParam> => {
  return defineAction({
    name: 'Tap',
    description: 'Tap the element',
    interfaceAlias: 'aiTap',
    paramSchema: actionTapParamSchema,
    call,
  });
};

// RightClick
const actionRightClickParamSchema = z.object({
  locate: getMidsceneLocationSchema().describe(
    'The element to be right clicked',
  ),
});
export type ActionRightClickParam = z.infer<typeof actionRightClickParamSchema>;

export const defineActionRightClick = (
  call: (param: ActionRightClickParam) => Promise<void>,
): DeviceAction<ActionRightClickParam> => {
  return defineAction({
    name: 'RightClick',
    description: 'Right click the element',
    interfaceAlias: 'aiRightClick',
    paramSchema: actionRightClickParamSchema,
    call,
  });
};

// Hover
const actionHoverParamSchema = z.object({
  locate: getMidsceneLocationSchema().describe('The element to be hovered'),
});
export type ActionHoverParam = z.infer<typeof actionHoverParamSchema>;

export const defineActionHover = (
  call: (param: ActionHoverParam) => Promise<void>,
): DeviceAction<ActionHoverParam> => {
  return defineAction({
    name: 'Hover',
    description: 'Move the mouse to the element',
    interfaceAlias: 'aiHover',
    paramSchema: actionHoverParamSchema,
    call,
  });
};

// Input
const actionInputParamSchema = z.object({
  value: z.string().describe('The value to be input'),
  locate: getMidsceneLocationSchema().describe('The element to be input'),
});
export type ActionInputParam = z.infer<typeof actionInputParamSchema>;

export const defineActionInput = (
  call: (param: ActionInputParam) => Promise<void>,
): DeviceAction<ActionInputParam> => {
  return defineAction({
    name: 'Input',
    description: 'Input the value into the element',
    interfaceAlias: 'aiInput',
    paramSchema: actionInputParamSchema,
    call,
  });
};

// KeyboardPress
const actionKeyboardPressParamSchema = z.object({
  locate: getMidsceneLocationSchema()
    .describe('The element to be clicked before pressing the key')
    .optional(),
  keyName: z.string().describe('The key to be pressed'),
});
export type ActionKeyboardPressParam = z.infer<
  typeof actionKeyboardPressParamSchema
>;

export const defineActionKeyboardPress = (
  call: (param: ActionKeyboardPressParam) => Promise<void>,
): DeviceAction<ActionKeyboardPressParam> => {
  return defineAction({
    name: 'KeyboardPress',
    description:
      'Press a function key, like "Enter", "Tab", "Escape". Do not use this to type text.',
    interfaceAlias: 'aiKeyboardPress',
    paramSchema: actionKeyboardPressParamSchema,
    call,
  });
};

// Scroll
const actionScrollParamSchema = z.object({
  direction: z
    .enum(['down', 'up', 'right', 'left'])
    .default('down')
    .describe('The direction to scroll'),
  scrollType: z
    .enum(['once', 'untilBottom', 'untilTop', 'untilRight', 'untilLeft'])
    .default('once')
    .describe('The scroll type'),
  distance: z
    .number()
    .nullable()
    .optional()
    .describe('The distance in pixels to scroll'),
  locate: getMidsceneLocationSchema()
    .optional()
    .describe('The element to be scrolled'),
});
export type ActionScrollParam = z.infer<typeof actionScrollParamSchema>;

export const defineActionScroll = (
  call: (param: ActionScrollParam) => Promise<void>,
): DeviceAction<ActionScrollParam> => {
  return defineAction({
    name: 'Scroll',
    description:
      'Scroll the page or an element. The direction to scroll, the scroll type, and the distance to scroll. The distance is the number of pixels to scroll. If not specified, use `down` direction, `once` scroll type, and `null` distance.',
    interfaceAlias: 'aiScroll',
    paramSchema: actionScrollParamSchema,
    call,
  });
};

// DragAndDrop
const actionDragAndDropParamSchema = z.object({
  from: getMidsceneLocationSchema().describe('The position to be dragged'),
  to: getMidsceneLocationSchema().describe('The position to be dropped'),
});
export type ActionDragAndDropParam = z.infer<
  typeof actionDragAndDropParamSchema
>;

export const defineActionDragAndDrop = (
  call: (param: ActionDragAndDropParam) => Promise<void>,
): DeviceAction<ActionDragAndDropParam> => {
  return defineAction({
    name: 'DragAndDrop',
    description: 'Drag and drop the element',
    interfaceAlias: 'aiDragAndDrop',
    paramSchema: actionDragAndDropParamSchema,
    call,
  });
};

export const commonWebActionsForWebPage = <T extends AbstractPage>(
  page: T,
): DeviceAction<any>[] => [
  defineActionTap(async (param) => {
    const element = param.locate;
    assert(element, 'Element not found, cannot tap');
    await page.mouse.click(element.center[0], element.center[1], {
      button: 'left',
    });
  }),
  defineActionRightClick(async (param) => {
    const element = param.locate;
    assert(element, 'Element not found, cannot right click');
    await page.mouse.click(element.center[0], element.center[1], {
      button: 'right',
    });
  }),
  defineActionHover(async (param) => {
    const element = param.locate;
    assert(element, 'Element not found, cannot hover');
    await page.mouse.move(element.center[0], element.center[1]);
  }),
  defineActionInput(async (param) => {
    const element = param.locate;
    if (element) {
      await page.clearInput(element as unknown as ElementInfo);

      if (!param || !param.value) {
        return;
      }
    }

    // Note: there is another implementation in AndroidDevicePage, which is more complex
    await page.keyboard.type(param.value);
  }),
  defineActionKeyboardPress(async (param) => {
    const element = param.locate;
    if (element) {
      await page.mouse.click(element.center[0], element.center[1], {
        button: 'left',
      });
    }

    const keys = getKeyCommands(param.keyName);
    await page.keyboard.press(keys as any); // TODO: fix this type error
  }),
  defineActionScroll(async (param) => {
    const element = param.locate;
    const startingPoint = element
      ? {
          left: element.center[0],
          top: element.center[1],
        }
      : undefined;
    const scrollToEventName = param?.scrollType;
    if (scrollToEventName === 'untilTop') {
      await page.scrollUntilTop(startingPoint);
    } else if (scrollToEventName === 'untilBottom') {
      await page.scrollUntilBottom(startingPoint);
    } else if (scrollToEventName === 'untilRight') {
      await page.scrollUntilRight(startingPoint);
    } else if (scrollToEventName === 'untilLeft') {
      await page.scrollUntilLeft(startingPoint);
    } else if (scrollToEventName === 'once' || !scrollToEventName) {
      if (param?.direction === 'down' || !param || !param.direction) {
        await page.scrollDown(param?.distance || undefined, startingPoint);
      } else if (param.direction === 'up') {
        await page.scrollUp(param.distance || undefined, startingPoint);
      } else if (param.direction === 'left') {
        await page.scrollLeft(param.distance || undefined, startingPoint);
      } else if (param.direction === 'right') {
        await page.scrollRight(param.distance || undefined, startingPoint);
      } else {
        throw new Error(`Unknown scroll direction: ${param.direction}`);
      }
      // until mouse event is done
      await sleep(500);
    } else {
      throw new Error(
        `Unknown scroll event type: ${scrollToEventName}, param: ${JSON.stringify(
          param,
        )}`,
      );
    }
  }),
  defineActionDragAndDrop(async (param) => {
    const from = param.from;
    const to = param.to;
    assert(from, 'missing "from" param for drag and drop');
    assert(to, 'missing "to" param for drag and drop');
    await page.mouse.drag(
      {
        x: from.center[0],
        y: from.center[1],
      },
      {
        x: to.center[0],
        y: to.center[1],
      },
    );
  }),
];
