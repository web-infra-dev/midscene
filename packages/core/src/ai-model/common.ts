import type { AIUsageInfo, Size } from '@/types';
import { assert } from '@midscene/shared/utils';

import type {
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
} from 'openai/resources';
import {
  callToGetJSONObject,
  checkAIConfig,
  getModelName,
} from './service-caller/index';

import type { PlanningLocateParam } from '@/types';

export type AIArgs = [
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
];

export enum AIActionType {
  ASSERT = 0,
  INSPECT_ELEMENT = 1,
  EXTRACT_DATA = 2,
  PLAN = 3,
}

export async function callAiFn<T>(
  msgs: AIArgs,
  AIActionTypeValue: AIActionType,
): Promise<{ content: T; usage?: AIUsageInfo }> {
  assert(
    checkAIConfig(),
    'Cannot find config for AI model service. If you are using a self-hosted model without validating the API key, please set `OPENAI_API_KEY` to any non-null value. https://midscenejs.com/model-provider.html',
  );

  const { content, usage } = await callToGetJSONObject<T>(
    msgs,
    AIActionTypeValue,
  );
  return { content, usage };
}

// transform the param of locate from qwen mode
export function fillLocateParam(locate: PlanningLocateParam) {
  if (locate?.bbox_2d && !locate?.bbox) {
    locate.bbox = locate.bbox_2d;
    // biome-ignore lint/performance/noDelete: <explanation>
    delete locate.bbox_2d;
  }

  const defaultBboxSize = 10;
  if (locate?.bbox) {
    locate.bbox[0] = Math.round(locate.bbox[0]);
    locate.bbox[1] = Math.round(locate.bbox[1]);
    locate.bbox[2] =
      typeof locate.bbox[2] === 'number'
        ? Math.round(locate.bbox[2])
        : Math.round(locate.bbox[0] + defaultBboxSize);
    locate.bbox[3] =
      typeof locate.bbox[3] === 'number'
        ? Math.round(locate.bbox[3])
        : Math.round(locate.bbox[1] + defaultBboxSize);
  }

  return locate;
}

let warned = false;
export function warnGPT4oSizeLimit(size: Size) {
  if (warned) return;
  if (getModelName()?.toLowerCase().includes('gpt-4o')) {
    const warningMsg = `GPT-4o has a maximum image input size of 2000x768 or 768x2000, but got ${size.width}x${size.height}. Please set your page to a smaller resolution. Otherwise, the result may be inaccurate.`;

    if (
      Math.max(size.width, size.height) > 2000 ||
      Math.min(size.width, size.height) > 768
    ) {
      console.warn(warningMsg);
      warned = true;
    }
  }
}
