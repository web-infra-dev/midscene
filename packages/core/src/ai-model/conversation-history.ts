import type { ChatCompletionMessageParam } from 'openai/resources/index';

export interface ConversationHistoryOptions {
  maxUserImageMessages?: number;
  initialMessages?: ChatCompletionMessageParam[];
}

export class ConversationHistory {
  private readonly maxUserImageMessages: number;
  private readonly messages: ChatCompletionMessageParam[] = [];

  constructor(options?: ConversationHistoryOptions) {
    this.maxUserImageMessages = options?.maxUserImageMessages ?? 4;
    if (options?.initialMessages?.length) {
      this.seed(options.initialMessages);
    }
  }

  append(message: ChatCompletionMessageParam) {
    if (message.role === 'user') {
      this.pruneOldestUserMessageIfNecessary();
    }

    this.messages.push(message);
  }

  seed(messages: ChatCompletionMessageParam[]) {
    this.reset();
    messages.forEach((message) => {
      this.append(message);
    });
  }

  reset() {
    this.messages.length = 0;
  }

  snapshot(): ChatCompletionMessageParam[] {
    return [...this.messages];
  }

  get length(): number {
    return this.messages.length;
  }

  [Symbol.iterator](): IterableIterator<ChatCompletionMessageParam> {
    return this.messages[Symbol.iterator]();
  }

  toJSON(): ChatCompletionMessageParam[] {
    return this.snapshot();
  }

  private pruneOldestUserMessageIfNecessary() {
    const userMessages = this.messages.filter((item) => item.role === 'user');
    if (userMessages.length < this.maxUserImageMessages) {
      return;
    }

    const firstUserMessageIndex = this.messages.findIndex(
      (item) => item.role === 'user',
    );

    if (firstUserMessageIndex >= 0) {
      this.messages.splice(firstUserMessageIndex, 1);
    }
  }
}
