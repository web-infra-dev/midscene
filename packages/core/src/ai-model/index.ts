export { callToGetJSONObject } from './service-caller/index';
export { systemPromptToLocateElement } from './prompt/llm-locator';
export { describeUserPage } from './prompt/util';

export type { ChatCompletionMessageParam } from 'openai/resources';

export {
  AiInspectElement,
  AiExtractElementInfo,
  AiAssert,
  transformElementPositionToId,
} from './inspect';

export { plan } from './llm-planning';
export { callAiFn } from './common';
export { vlmPlanning } from './ui-tars-planning';
