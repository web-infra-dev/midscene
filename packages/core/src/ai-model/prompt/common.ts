import type { vlLocateMode } from '@midscene/shared/env';
export function bboxDescription(vlMode: ReturnType<typeof vlLocateMode>) {
  if (vlMode === 'gemini') {
    return '2d bounding box as [ymin, xmin, ymax, xmax]';
  }
  return '2d bounding box as [xmin, ymin, xmax, ymax]';
}
