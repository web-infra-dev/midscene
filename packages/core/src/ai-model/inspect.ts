import {
  MIDSCENE_USE_QWEN_VL,
  MIDSCENE_USE_VLM_UI_TARS,
  getAIConfigInBoolean,
} from '@/env';
import type {
  AIAssertionResponse,
  AIElementIdResponse,
  AIElementResponse,
  AISectionParseResponse,
  AISingleElementResponse,
  AISingleElementResponseByPosition,
  AIUsageInfo,
  BaseElement,
  ElementById,
  ElementTreeNode,
  Size,
  UIContext,
} from '@/types';
import { paddingToMatchBlock } from '@midscene/shared/img';
import { assert } from '@midscene/shared/utils';
import type {
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
} from 'openai/resources';
import { AIActionType, callAiFn } from './common';
import { systemPromptToAssert } from './prompt/assertion';
import { extractDataPrompt, systemPromptToExtract } from './prompt/extraction';
import {
  findElementPrompt,
  systemPromptToLocateElement,
} from './prompt/llm-locator';
import {
  describeUserPage,
  distance,
  distanceThreshold,
  elementByPositionWithElementInfo,
} from './prompt/util';
import { callToGetJSONObject } from './service-caller/index';

export type AIArgs = [
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
];

const liteContextConfig = {
  filterNonTextContent: true,
  truncateTextLength: 200,
};

function transformToAbsoluteCoords(
  relativePosition: { x: number; y: number },
  size: Size,
) {
  return {
    x: Number(((relativePosition.x / 1000) * size.width).toFixed(3)),
    y: Number(((relativePosition.y / 1000) * size.height).toFixed(3)),
  };
}

// let index = 0;
export async function transformElementPositionToId(
  aiResult: AIElementResponse | [number, number],
  treeRoot: ElementTreeNode<BaseElement>,
  size: { width: number; height: number },
  insertElementByPosition: (position: { x: number; y: number }) => BaseElement,
) {
  const emptyResponse: AIElementResponse = {
    errors: [],
    elements: [],
  };

  const elementAtPosition = (center: { x: number; y: number }) => {
    const element = elementByPositionWithElementInfo(treeRoot, center);
    const distanceToCenter = element
      ? distance({ x: element.center[0], y: element.center[1] }, center)
      : 0;
    return distanceToCenter <= distanceThreshold ? element : undefined;
  };

  if ('bbox' in aiResult) {
    if (
      !Array.isArray(aiResult.bbox) ||
      (aiResult.bbox as number[]).length !== 4
    ) {
      return emptyResponse;
    }

    aiResult.bbox[0] = Math.ceil(aiResult.bbox[0]);
    aiResult.bbox[1] = Math.ceil(aiResult.bbox[1]);
    aiResult.bbox[2] = Math.ceil(aiResult.bbox[2]);
    aiResult.bbox[3] = Math.ceil(aiResult.bbox[3]);

    const centerX = (aiResult.bbox[0] + aiResult.bbox[2]) / 2;
    const centerY = (aiResult.bbox[1] + aiResult.bbox[3]) / 2;

    let element = elementAtPosition({ x: centerX, y: centerY });

    if (!element) {
      element = insertElementByPosition({
        x: centerX,
        y: centerY,
      });
    }
    assert(
      element,
      `inspect: no element found with coordinates: ${JSON.stringify(aiResult.bbox)}`,
    );
    return {
      errors: [],
      elements: [
        {
          id: element.id,
        },
      ],
    };
  }

  if (Array.isArray(aiResult)) {
    // [number, number] coord
    const relativePosition = aiResult;
    const absolutePosition = transformToAbsoluteCoords(
      {
        x: relativePosition[0],
        y: relativePosition[1],
      },
      size,
    );

    let element = elementAtPosition(absolutePosition);
    if (!element) {
      element = insertElementByPosition(absolutePosition);
    }

    assert(
      element,
      `inspect: no id found with position: ${JSON.stringify({ absolutePosition })}`,
    );

    return {
      errors: [],
      elements: [
        {
          id: element.id,
        },
      ],
    };
  }

  return {
    errors: aiResult.errors,
    elements: aiResult.elements,
  };
}

function matchQuickAnswer(
  quickAnswer:
    | Partial<AISingleElementResponse>
    | Partial<AISingleElementResponseByPosition>
    | undefined,
  tree: ElementTreeNode<BaseElement>,
  elementById: ElementById,
  insertElementByPosition: (position: { x: number; y: number }) => BaseElement,
): Awaited<ReturnType<typeof AiInspectElement>> | undefined {
  if (!quickAnswer) {
    return undefined;
  }
  if ('id' in quickAnswer && quickAnswer.id && elementById(quickAnswer.id)) {
    return {
      parseResult: {
        elements: [quickAnswer as AISingleElementResponse],
        errors: [],
      },
      rawResponse: quickAnswer,
      elementById,
    };
  }

  if ('position' in quickAnswer && quickAnswer.position) {
    let element = elementByPositionWithElementInfo(tree, quickAnswer.position);
    if (!element) {
      element = insertElementByPosition(quickAnswer.position);
    }
    return {
      parseResult: {
        elements: [element],
        errors: [],
      },
      rawResponse: quickAnswer,
      elementById,
    } as any;
  }

  if ('bbox' in quickAnswer && quickAnswer.bbox) {
    const centerPosition = {
      x: Math.floor((quickAnswer.bbox[0] + quickAnswer.bbox[2]) / 2),
      y: Math.floor((quickAnswer.bbox[1] + quickAnswer.bbox[3]) / 2),
    };
    let element = elementByPositionWithElementInfo(tree, centerPosition);
    if (!element) {
      element = insertElementByPosition(centerPosition);
    }
    return {
      parseResult: {
        elements: [element],
        errors: [],
      },
      rawResponse: quickAnswer,
      elementById,
    } as any;
  }

  return undefined;
}

export async function AiInspectElement<
  ElementType extends BaseElement = BaseElement,
>(options: {
  context: UIContext<ElementType>;
  targetElementDescription: string;
  callAI?: typeof callAiFn<AIElementResponse | [number, number]>;
  quickAnswer?: Partial<
    AISingleElementResponse | AISingleElementResponseByPosition
  >;
}): Promise<{
  parseResult: AIElementIdResponse;
  rawResponse: any;
  elementById: ElementById;
  usage?: AIUsageInfo;
}> {
  const { context, targetElementDescription, callAI } = options;
  const { screenshotBase64, screenshotBase64WithElementMarker } = context;
  const { description, elementById, insertElementByPosition, size } =
    await describeUserPage(context);
  // meet quick answer
  const quickAnswer = matchQuickAnswer(
    options.quickAnswer,
    context.tree,
    elementById,
    insertElementByPosition,
  );
  if (quickAnswer) {
    return quickAnswer;
  }

  assert(
    targetElementDescription,
    'cannot find the target element description',
  );

  const userInstructionPrompt = await findElementPrompt.format({
    pageDescription: description,
    targetElementDescription,
  });
  const systemPrompt = systemPromptToLocateElement();

  let imagePayload = screenshotBase64WithElementMarker || screenshotBase64;

  if (getAIConfigInBoolean(MIDSCENE_USE_QWEN_VL)) {
    imagePayload = await paddingToMatchBlock(imagePayload);
  }

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: [
        {
          type: 'image_url',
          image_url: {
            url: imagePayload,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: userInstructionPrompt,
        },
      ],
    },
  ];

  const callAIFn =
    callAI || callToGetJSONObject<AIElementResponse | [number, number]>;

  const res = await callAIFn(msgs, AIActionType.INSPECT_ELEMENT);

  const parseResult = await transformElementPositionToId(
    res.content,
    context.tree,
    size,
    insertElementByPosition,
  );

  return {
    parseResult,
    rawResponse: res.content,
    elementById,
    usage: res.usage,
  };
}

export async function AiExtractElementInfo<
  T,
  ElementType extends BaseElement = BaseElement,
>(options: {
  dataQuery: string | Record<string, string>;
  context: UIContext<ElementType>;
}) {
  const { dataQuery, context } = options;
  const systemPrompt = systemPromptToExtract();

  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(
    context,
    liteContextConfig,
  );

  let dataKeys = '';
  let dataQueryText = '';
  if (typeof dataQuery === 'string') {
    dataKeys = '';
    dataQueryText = dataQuery;
  } else {
    dataKeys = `return in key-value style object, keys are ${Object.keys(dataQuery).join(',')}`;
    dataQueryText = JSON.stringify(dataQuery, null, 2);
  }
  const extractDataPromptText = await extractDataPrompt.format({
    pageDescription: description,
    dataKeys,
    dataQuery: dataQueryText,
  });

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: [
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: extractDataPromptText,
        },
      ],
    },
  ];

  const result = await callAiFn<AISectionParseResponse<T>>(
    msgs,
    AIActionType.EXTRACT_DATA,
  );
  return {
    parseResult: result.content,
    elementById,
    usage: result.usage,
  };
}

export async function AiAssert<
  ElementType extends BaseElement = BaseElement,
>(options: { assertion: string; context: UIContext<ElementType> }) {
  const { assertion, context } = options;

  assert(assertion, 'assertion should be a string');

  const { screenshotBase64 } = context;

  const systemPrompt = systemPromptToAssert({
    isUITars: getAIConfigInBoolean(MIDSCENE_USE_VLM_UI_TARS),
  });

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: [
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: `
Here is the assertion. Please tell whether it is truthy according to the screenshot.
=====================================
${assertion}
=====================================
  `,
        },
      ],
    },
  ];

  const { content: assertResult, usage } = await callAiFn<AIAssertionResponse>(
    msgs,
    AIActionType.ASSERT,
  );
  return {
    content: assertResult,
    usage,
  };
}
