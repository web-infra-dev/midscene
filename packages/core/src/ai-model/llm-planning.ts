import { MIDSCENE_USE_QWEN_VL, getAIConfigInBoolean } from '@/env';
import type { PlanningAIResponse, UIContext } from '@/types';
import { paddingToMatchBlock } from '@midscene/shared/img';
import { assert } from '@midscene/shared/utils';
import {
  AIActionType,
  type AIArgs,
  callAiFn,
  fillLocateParam,
  warnGPT4oSizeLimit,
} from './common';
import {
  automationUserPrompt,
  generateTaskBackgroundContext,
  systemPromptToTaskPlanning,
} from './prompt/llm-planning';
import { describeUserPage } from './prompt/util';

export async function plan(
  userInstruction: string,
  opts: {
    log?: string;
    context: UIContext;
    callAI?: typeof callAiFn<PlanningAIResponse>;
  },
): Promise<PlanningAIResponse> {
  const { callAI, context } = opts || {};
  const { screenshotBase64, screenshotBase64WithElementMarker, size } = context;
  const { description: pageDescription } = await describeUserPage(context);

  const systemPrompt = await systemPromptToTaskPlanning();
  const taskBackgroundContextText = generateTaskBackgroundContext(
    userInstruction,
    opts.log,
  );
  const userInstructionPrompt = await automationUserPrompt().format({
    pageDescription,
    taskBackgroundContext: taskBackgroundContextText,
  });

  let imagePayload = screenshotBase64WithElementMarker || screenshotBase64;
  if (getAIConfigInBoolean(MIDSCENE_USE_QWEN_VL)) {
    imagePayload = await paddingToMatchBlock(imagePayload);
  }

  warnGPT4oSizeLimit(size);

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: [
        {
          type: 'image_url',
          image_url: {
            url: imagePayload,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: userInstructionPrompt,
        },
      ],
    },
  ];

  const call = callAI || callAiFn;
  const { content, usage } = await call(msgs, AIActionType.PLAN);
  const rawResponse = JSON.stringify(content, undefined, 2);
  const planFromAI = content;

  const actions =
    (planFromAI.action?.type ? [planFromAI.action] : planFromAI.actions) || [];
  const returnValue: PlanningAIResponse = {
    ...planFromAI,
    actions,
    rawResponse,
    usage,
  };

  assert(planFromAI, "can't get plans from AI");

  if (getAIConfigInBoolean(MIDSCENE_USE_QWEN_VL)) {
    actions.forEach((action) => {
      if (action.locate) {
        action.locate = fillLocateParam(action.locate);
      }
    });
    // in Qwen-VL, error means error. In GPT-4o, error may mean more actions are needed.
    assert(!planFromAI.error, `Failed to plan actions: ${planFromAI.error}`);
  }

  if (
    actions.length === 0 &&
    returnValue.more_actions_needed_by_instruction &&
    !returnValue.sleep
  ) {
    console.warn(
      'No actions planned for the prompt, but model said more actions are needed:',
      userInstruction,
    );
  }

  return returnValue;
}
