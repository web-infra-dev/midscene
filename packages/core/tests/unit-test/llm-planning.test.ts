import { fillLocateParam } from '@/ai-model/common';
import { afterEach, beforeEach, describe, expect, it } from 'vitest';

describe('llm planning - qwen', () => {
  let originalMidsceneUseQwenVl: string | undefined;
  let originalMidsceneUseDoubaoVl: string | undefined;
  beforeEach(() => {
    originalMidsceneUseQwenVl = process.env.MIDSCENE_USE_QWEN_VL;
    originalMidsceneUseDoubaoVl = process.env.MIDSCENE_USE_DOUBAO_VISION;
    process.env.MIDSCENE_USE_QWEN_VL = 'true';
    process.env.MIDSCENE_USE_DOUBAO_VISION = 'false';
  });

  afterEach(() => {
    process.env.MIDSCENE_USE_QWEN_VL = originalMidsceneUseQwenVl;
    process.env.MIDSCENE_USE_DOUBAO_VISION = originalMidsceneUseDoubaoVl;
  });

  it('fill locate param', () => {
    const locate = {
      id: 'test',
      prompt: 'test',
      bbox_2d: [100, 100, 200, 200] as [number, number, number, number],
    };

    const filledLocate = fillLocateParam(locate, 1000, 1000);
    expect(filledLocate).toEqual({
      id: 'test',
      prompt: 'test',
      bbox: [100, 100, 200, 200],
    });
  });

  it('fill locate param', () => {
    const locate = {
      id: 'test',
      prompt: 'test',
      bbox_2d: [100, 100] as unknown as [number, number, number, number],
    };

    const filledLocate = fillLocateParam(locate, 1000, 1000);
    expect(filledLocate).toEqual({
      id: 'test',
      prompt: 'test',
      bbox: [100, 100, 120, 120],
    });
  });
});

describe('llm planning - doubao', () => {
  let originalMidsceneUseDoubaoVl: string | undefined;
  let originalMidsceneUseQwenVl: string | undefined;
  beforeEach(() => {
    originalMidsceneUseDoubaoVl = process.env.MIDSCENE_USE_DOUBAO_VISION;
    originalMidsceneUseQwenVl = process.env.MIDSCENE_USE_QWEN_VL;
    process.env.MIDSCENE_USE_DOUBAO_VISION = 'true';
    process.env.MIDSCENE_USE_QWEN_VL = 'false';
  });

  afterEach(() => {
    process.env.MIDSCENE_USE_DOUBAO_VISION = originalMidsceneUseDoubaoVl;
    process.env.MIDSCENE_USE_QWEN_VL = originalMidsceneUseQwenVl;
  });

  it('fill locate param', () => {
    const locate = {
      id: 'test',
      prompt: 'test',
      bbox_2d: [923, 123, 123, 123] as [number, number, number, number],
    };

    const filledLocate = fillLocateParam(locate, 1000, 1000);
    expect(filledLocate).toEqual({
      id: 'test',
      prompt: 'test',
      bbox: [923, 123, 123, 123],
    });
  });
});
