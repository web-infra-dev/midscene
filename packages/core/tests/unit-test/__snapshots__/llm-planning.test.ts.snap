// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`llm planning - build yaml flow > build yaml flow 1`] = `
[
  {
    "aiInput": "hello",
    "locate": "The input box for adding a new todo",
  },
  {
    "aiHover": "The second item 'Learn Rust' in the task list",
  },
  {
    "aiTap": "The input box labeled 'What needs to be done?'",
  },
  {
    "aiScroll": null,
    "direction": "down",
    "distance": 500,
    "locate": "some button",
    "scrollType": "once",
  },
]
`;
